lib/irslackd.js

```diff
@@ -299,13 +310,121 @@ class Irslackd {
+        return; // must have two params (nick and channel)
+      }
+      var userId = await self.lookupUserId(ircUser, msg.args[0].substr(0, 1) === '@' ? msg.args[0].substr(1) : msg.args[0]); // remove at-sign
+      var channelId = await self.lookupChannelId(ircUser, msg.args[1].substr(0, 1) === '#' ? msg.args[1].substr(1) : msg.args[1]); // remove hash
+      var result = await ircUser.slackWeb.paginateCallOrThrow('conversations.invite', 'channel', {
+        users: userId,
+        channel: channelId,
+      });
+      console.debug('invite call response: ' + JSON.stringify(result));
+      self.ircd.write(ircUser.socket, 'irslackd', '341', [
+        ircUser.ircNick,
+        userId,
+        channelId,
+      ]);
+    } catch (e) {
+      var errorFromData = typeof JSON.parse(e.message).error.data !== 'undefined' ? JSON.parse(e.message).error.data.error : 'no data element in error response';
+      self.ircd.write(ircUser.socket, 'irslackd', '371', [
+        ircUser.ircNick, ':/INVITE failed->' +
+        JSON.parse(e.message).error.code + '->' +
+        errorFromData,
+      ]);
+    }
+  }
+  async lookupChannelId(ircUser, channelName) {
+    var channelId;
+    let conversations = await ircUser.slackWeb.paginateCallOrThrow('conversations.list', 'channels', {
+      types: 'public_channel, private_channel, mpim, im',
+      exclude_archived: true,
+      limit: 1000,
+    });
+    var BreakException = {};
+    try {
+      conversations.channels.forEach((channel) => {
+        if (channel.name === channelName) {
+          channelId = channel.id;
+          throw BreakException;
+        }
+      });
+    } catch (e) {
+      if (e !== BreakException) throw e;
+    }
+    return channelId;
+  }
+  async lookupUserId(ircUser, userName) {
+    var userId;
+    let users = await ircUser.slackWeb.paginateCallOrThrow('users.list', 'members', {
+      presence: false,
+      limit: 1000,
+    });
+    var BreakException = {};
+    try {
+      users.members.forEach((member) => {
+        if (member.name === userName) {
+          userId = member.id;
+          throw BreakException;
+        }
+      });
+    } catch (e) {
+      if (e !== BreakException) throw e;
     }
+    return userId;
+  }
+  async onIrcWho(ircUser, msg) {
+    const self = this;
+    var userSearch;
+    if (msg.args.length > 1) return; // Only support `WHO` with one params
+    if (msg.args.length > 0 && msg.args[0].substr(0, 1) === '#') {
+      self.ircd.write(ircUser.socket, 'irslackd', '315', [ircUser.ircNick, ':End of WHO - channels not implemented']);
+      return;
+    }
+    if (msg.args.length > 0) {
+      userSearch = msg.args[0];
+    }
+    self.ircd.write(ircUser.socket, 'irslackd', '371', [ircUser.ircNick, ':Start of WHO']);
+    let users = await ircUser.slackWeb.paginateCallOrThrow('users.list', 'members', {
+      presence: false,
+      limit: 1000,
+    });
```

Comment:
`users.list` can take a long time for large Slack workspaces with tens of thousands of users. I suspect it will be more efficient in most cases to string match on entries in `ircToSlack` and then call `users.info` for each match. Each `users.info` call can run in parallel (no await), and we can use `Promise.race` to send `352` replies as each call finishes.

Reply:
yeah, I didn't realize you were keeping the ircToSlack map ... i'll look at that instead